bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py

```diff
@@ -0,0 +1,509 @@
+        self.res_ptr_id += 1
+
+    def _get_template_args(self, operation):
+        '''
+        Gets the template arguments with which to generate a given operation.
+
+        Args:
+            operation (Operation): object representing the operation whose
+                template arguments need to be returned.
+
+        Returns:
+            string: template arguments for this operation.
+        '''
+
+        # TODO: generate templated operations when possible, e.g. Max<double>
+
+        return ''
+
+    def _process_args(self, operation):
+        '''
+        Processes the arguments of an operation node.
+
+        Args:
+            operation (Operation): object representing the operation whose
+                arguments need to be processed.
+        '''
+
+        # TODO
+        # - Do a more thorough type conversion
+        # - Use RDF helper functions to convert jitted strings to lambdas
+
+        # Argument type conversion
+        for narg, arg in enumerate(operation.args):
+            if (narg > 0):
+                self.graph_nodes += ', '
+
+            if isinstance(arg, str):
+                self.graph_nodes += '"{}"'.format(arg)
+            elif isinstance(arg, tuple):
+                self.graph_nodes += '{'
+                for nelem, elem in enumerate(arg):
+                    if nelem > 0:
+                        self.graph_nodes += ','
+                    if isinstance(elem, str):
+                        self.graph_nodes += '"{}"'.format(elem)
+                    else:
+                        self.graph_nodes += '{}'.format(elem)
+                self.graph_nodes += '}'
+
+        # Make Snapshot lazy
+        # TODO: Do a proper processing of the args (user might have specified
+        # her own options object)
+        if operation.name == 'Snapshot':
+            self.graph_nodes += ', "", lazy_options'
+
+
+    def execute(self, rdf):
+        '''
+        Compiles the workflow generation code and executes it.
+
+        Args:
+            rdf (ROOT::RDF::RNode): object that represents the dataset on
+                which to execute the workflow.
+
+        Returns:
+            tuple: the first element is the list of results of the actions in
+                the C++ workflow, the second element is the list of result types
+                corresponding to those actions.
+        '''
+
+        function_id = self._compile()
+        return self._run_function(rdf, function_id)
+
+    def _compile(self):
+        '''
+        Generates the workflow code C++ file and compiles it with ACLiC
+        into a shared library. The library is also loaded as part of the
+        `TSystem::CompileMacro` call.
+
+        The name of the generated C++ file contains both a hash of its
+        code and the ID of the process that created it. This is done to
+        prevent clashes between multiple (non-sandboxed) worker processes
+        that try to write to the same file concurrently.
```

Comment:
This is very nice and true for multiprocessing, not sure about multithreading. I'd say we need at least:
* a lock on the file being written
* the first thread arriving to this function needs to compile the function before other threads so it can fill the cache of the class 

Although by looking at the requirements above it seems to me this mechanism would need to happen just once per computation graph per worker ? First process/thread that starts this compilation locks a file with the name being just the hash of the code so all other processes/threads would be waiting for that file to unlock? Not sure

Reply:
As you can see I have a ` TODO: Make this function thread-safe? To support Dask threaded workers` at the beginning of this function :smile: 

Right now the cache is just reused by the same worker process that wants to run the same workflow on another fragment of the dataset, in the same application run. It could also be the case, if the worker process survives across different application runs, that it reuses a cache from a previous run (because the user submitted a second job with the exact same workflow).

If we want to support thread-safe caching, we need to have a look at the alternatives for locking a file from Python. It seems there are a few packages that provide advisory locking features (which is good enough for us), you can have a read at [this thread](https://stackoverflow.com/questions/489861/locking-a-file-in-python) for a discussion on this topic.

However, I would leave this for a second round, since testing the caching is not what has highest priority now: it is testing the performance of the optimized workflow generation vs the old way (with new jitting optimizations).

Reply:
Yes I saw your comment, I just wanted to start a conversation on this because I think it is a very interesting part of this feature :smile: I agree it can be left for a second iteration :+1: